docs/antrea-ipam.md

```diff
@@ -46,3 +46,147 @@ cluster. Valid range is 16 to 30. Default is 24.
 
 - `nodeCIDRMaskSizeIPv6`: Mask size for IPv6 Node CIDR in IPv6 or dual-stack
 cluster. Valid range is 64 to 126. Default is 64.
+
+## Antrea Flexible-IPAM
+
+Antrea supports flexible control over Pod IP addressing since version 1.4. This can be
+achieved by configuring `IPPool` CRD with a desired set of IP ranges. The `IPPool` can be
+annotated to Namespace, Pod and PodTemplate of StatefulSet/Deployment. Antrea will manage
+IP address assignment for corresponding Pods according to `IPPool` spec.
+
+Note that the IP pool annotation cannot be updated or deleted without recreating the
+resource. An `IPPool` can be extended, but cannot be shrunk if already assigned to a
+resource. The IP ranges of IPPools must not overlap, otherwise it would lead to undefined
+behavior.
+
+Regular `Subnet per Node` IPAM will continue to be used for resources without the IP pool
+annotation, or when `AntreaIPAM` feature is disabled.
+
+### Usage
+
+#### Enable AntreaIPAM feature gate
+
+Users need to enable `AntreaIPAM` from the featureGates map defined in antrea.yml for
+both Controller and Agent:
+
+```yaml
+  antrea-controller.conf: |
+    ...
+    featureGates:
+    # Enable flexible IPAM mode for Antrea. This mode allows to assign IP Ranges to Namespaces,
+    # Deployments and StatefulSets via IP Pool annotation.
+      AntreaIPAM: true
+    ...
+  antrea-agent.conf: |
+    ...
+    featureGates:
+    # Enable flexible IPAM mode for Antrea. This mode allows to assign IP Ranges to Namespaces,
+    # Deployments and StatefulSets via IP Pool annotation.
+      AntreaIPAM: true
+    ...
+```
+
+#### Create IPPool CR
+
+The following example YAML manifests create an IPPool CR.
+
+```yaml
+apiVersion: "crd.antrea.io/v1alpha2"
+kind: IPPool
+metadata:
+  name: pool1
+spec:
+  ipVersion: 4
+  ipRanges:
+  - start: "10.2.0.12"
+    end: "10.2.0.20"
+    gateway: "10.2.0.1"
+    prefixLength: 24
+```
+
+#### IPPool Annotations on Namespace
+
+The following example YAML manifests create a Namespace to allocate Pod IPs from the IP pool.
+
+```yaml
+kind: Namespace
+metadata:
+  annotations:
+    ipam.antrea.io/ippools: 'pool1'
+...
+```
+
+#### IPPool Annotations on Pod (available since Antrea 1.5)
+
+Since Antrea 1.5, Pod IPPool annotation is supported and has a higher priority than the
+Namespace IPPool annotation. This annotation can be added to `PodTemplate` of a
+controller resource such as StatefulSet and Deployment.
+
+Pod IP annotation is supported for a single Pod to specify a fixed IP for the Pod.
+
+Examples of annotations on a Pod or PodTemplate:
+
+```yaml
+kind: StatefulSet
+spec:
+  template:
+    metadata:
+      annotations:
+        ipam.antrea.io/ippools: 'sts-ip-pool1'  # This annotation will be set automatically on all Pods managed by this resource
+        # Do not add pod-ips annotation to PodTemplate
```

Comment:
Would the following be more accurate?
```suggestion
        # Do not add pod-ips annotation to PodTemplate if there is more than 1 replica
```

Reply:
I think it is accurate, but I also feel not to confuse user to add `pod-ips` annotation into PodTemplate. If user increases replicas after StatefulSet with `pod-ips` PodTemplate annotation created, the result will be wrong.
Thus I prefer to notify user not to add `pod-ips` annotation to PodTemplate. For your use case, user should create a dedicated IPPool which contains >=1 free IP(s) for this StatefulSet.

Reply:
But there is no such examples in this document yet. And if only 1 replica is desired for a StatefulSet or Deployment, it should be easier to just reuse the ippool to get a fixed IP?
Isn't the condition "if there is more than 1 replica" already clear that user shouldn't increase replica?

Reply:
Got it. Added example for replicas=1 and replicas>1.